// @meshconnect/web-link-sdk/Link.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype);
    return g.next = verb(0), g["throw"] = verb(1), g["return"] = verb(2), typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { addPopup, iframeId, removePopup } from './utils/popup';
import { isLinkEventTypeKey } from './utils/event-types';
import { sdkSpecs } from './utils/sdk-specs';
import { BridgeParent } from '@meshconnect/uwc-bridge-parent';
var currentOptions;
var targetOrigin;
var linkTokenOrigin;
var currentIframeId = iframeId;
var bridgeParent = null;
var iframeElement = function () {
    return document.getElementById(currentIframeId);
};
function sendMessageToIframe(message) {
    var _a;
    var iframe = iframeElement();
    if (!iframe) {
        console.warn("Mesh SDK: Failed to deliver ".concat(message.type, " message to the iframe - no iframe element found"));
        return;
    }
    if (!linkTokenOrigin) {
        console.warn("Mesh SDK: Failed to deliver ".concat(message.type, " message to the iframe - no link token origin found"));
        return;
    }
    try {
        (_a = iframe.contentWindow) === null || _a === void 0 ? void 0 : _a.postMessage(message, linkTokenOrigin);
    }
    catch (e) {
        console.error("Mesh SDK: Failed to deliver ".concat(message.type, " message to the iframe"));
        console.error(e);
    }
}
function handleLinkEvent(event) {
    return __awaiter(this, void 0, void 0, function () {
        var payload, payload, payload, payload;
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        return __generator(this, function (_l) {
            switch (event.data.type) {
                case 'brokerageAccountAccessToken': {
                    payload = {
                        accessToken: event.data.payload
                    };
                    (_a = currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.onEvent) === null || _a === void 0 ? void 0 : _a.call(currentOptions, {
                        type: 'integrationConnected',
                        payload: payload
                    });
                    (_b = currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.onIntegrationConnected) === null || _b === void 0 ? void 0 : _b.call(currentOptions, payload);
                    break;
                }
                case 'delayedAuthentication': {
                    payload = {
                        delayedAuth: event.data.payload
                    };
                    (_c = currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.onEvent) === null || _c === void 0 ? void 0 : _c.call(currentOptions, {
                        type: 'integrationConnected',
                        payload: payload
                    });
                    (_d = currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.onIntegrationConnected) === null || _d === void 0 ? void 0 : _d.call(currentOptions, payload);
                    break;
                }
                case 'transferFinished': {
                    payload = event.data.payload;
                    (_e = currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.onEvent) === null || _e === void 0 ? void 0 : _e.call(currentOptions, {
                        type: 'transferCompleted',
                        payload: payload
                    });
                    (_f = currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.onTransferFinished) === null || _f === void 0 ? void 0 : _f.call(currentOptions, payload);
                    break;
                }
                case 'close':
                case 'done': {
                    payload = (_g = event.data) === null || _g === void 0 ? void 0 : _g.payload;
                    (_h = currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.onExit) === null || _h === void 0 ? void 0 : _h.call(currentOptions, payload === null || payload === void 0 ? void 0 : payload.errorMessage, payload);
                    bridgeParent === null || bridgeParent === void 0 ? void 0 : bridgeParent.destroy();
                    removePopup();
                    break;
                }
                case 'loaded': {
                    sendMessageToIframe({
                        type: 'meshSDKSpecs',
                        payload: __assign({}, sdkSpecs)
                    });
                    if (currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.accessTokens) {
                        sendMessageToIframe({
                            type: 'frontAccessTokens',
                            payload: currentOptions.accessTokens
                        });
                    }
                    if (currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.transferDestinationTokens) {
                        sendMessageToIframe({
                            type: 'frontTransferDestinationTokens',
                            payload: currentOptions.transferDestinationTokens
                        });
                    }
                    (_j = currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.onEvent) === null || _j === void 0 ? void 0 : _j.call(currentOptions, { type: 'pageLoaded' });
                    break;
                }
                default: {
                    if (isLinkEventTypeKey(event.data.type)) {
                        (_k = currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.onEvent) === null || _k === void 0 ? void 0 : _k.call(currentOptions, event.data);
                    }
                    break;
                }
            }
            return [2 /*return*/];
        });
    });
}
function eventsListener(event) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!(event.origin !== targetOrigin && event.origin !== linkTokenOrigin)) return [3 /*break*/, 1];
                    console.warn('Received message from untrusted origin:', event.origin);
                    return [3 /*break*/, 3];
                case 1: return [4 /*yield*/, handleLinkEvent(event)];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    });
}
export var createLink = function (options) {
    var openLink = function (linkToken, customIframeId) {
        var _a, _b;
        if (!linkToken) {
            (_a = options === null || options === void 0 ? void 0 : options.onExit) === null || _a === void 0 ? void 0 : _a.call(options, 'Invalid link token!');
            return;
        }
        currentOptions = options;
        var linkUrl = window.atob(linkToken);
        var isProtocolValid = linkUrl.startsWith('http://') || linkUrl.startsWith('https://');
        if (!isProtocolValid) {
            (_b = options === null || options === void 0 ? void 0 : options.onExit) === null || _b === void 0 ? void 0 : _b.call(options, 'Invalid link token!');
            return;
        }
        linkUrl = addLanguage(linkUrl, currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.language);
        linkUrl = addDisplayFiatCurrency(linkUrl, currentOptions === null || currentOptions === void 0 ? void 0 : currentOptions.displayFiatCurrency);
        linkTokenOrigin = new URL(linkUrl).origin;
        window.removeEventListener('message', eventsListener);
        if (customIframeId) {
            var iframe_1 = document.getElementById(customIframeId);
            if (iframe_1) {
                iframe_1.allow = 'clipboard-read *; clipboard-write *';
                iframe_1.src = linkUrl;
                currentIframeId = customIframeId;
            }
            else {
                console.warn("Mesh SDK: No iframe found with id ".concat(customIframeId));
            }
        }
        else {
            currentIframeId = iframeId;
            addPopup(linkUrl);
        }
        window.addEventListener('message', eventsListener);
        targetOrigin = window.location.origin;
        var iframe = iframeElement();
        if (iframe) {
            bridgeParent = new BridgeParent(iframe);
        }
    };
    var closeLink = function () {
        var _a;
        bridgeParent === null || bridgeParent === void 0 ? void 0 : bridgeParent.destroy();
        removePopup();
        window.removeEventListener('message', eventsListener);
        (_a = options.onExit) === null || _a === void 0 ? void 0 : _a.call(options);
    };
    return {
        openLink: openLink,
        closeLink: closeLink
    };
};
function addLanguage(linkUrl, language) {
    if (language === 'system') {
        language =
            typeof navigator !== 'undefined' && navigator.language
                ? encodeURIComponent(navigator.language)
                : undefined;
    }
    return "".concat(linkUrl).concat(linkUrl.includes('?') ? '&' : '?', "lng=").concat(language || 'en');
}
function addDisplayFiatCurrency(linkUrl, displayFiatCurrency) {
    if (displayFiatCurrency) {
        return "".concat(linkUrl).concat(linkUrl.includes('?') ? '&' : '?', "fiatCur=").concat(displayFiatCurrency);
    }
    return linkUrl;
}
