// @meshconnect/uwc-bridge-parent/dist/bridge-parent.js
import { getWallets } from '@wallet-standard/app';
import { StandardWalletAdapter } from './standard-wallet-adapter';
export class BridgeParent {
    detectedEip6963Wallets = new Map();
    detectedWalletStandardWallets = new Map();
    walletStandardUnsubscribe;
    constructor() {
        this.initialize();
    }
    initialize() {
        this.detectEIP6963Wallets();
        this.detectWalletStandardWallets();
        this.logAllWallets();
    }
    detectEIP6963Wallets() {
        if (typeof window === 'undefined') {
            console.warn('BridgeParent: Window is not available, skipping EIP-6963 detection');
            return;
        }
        const handleAnnouncement = (event) => {
            const { info, provider } = event.detail;
            const detectedWallet = {
                info,
                provider
            };
            this.detectedEip6963Wallets.set(info.uuid, detectedWallet);
            console.log(`EIP-6963 wallet detected:`, info, provider);
        };
        window.addEventListener('eip6963:announceProvider', handleAnnouncement);
        window.dispatchEvent(new Event('eip6963:requestProvider'));
        setTimeout(() => {
            window.removeEventListener('eip6963:announceProvider', handleAnnouncement);
        }, 3000);
    }
    detectWalletStandardWallets() {
        if (typeof window === 'undefined') {
            console.warn('BridgeParent: Window is not available, skipping Wallet Standard detection');
            return;
        }
        try {
            const { get, on } = getWallets();
            const handleWallet = (wallet) => {
                // Create adapter for the wallet
                const adapter = new StandardWalletAdapter(wallet);
                // Only add Solana wallets
                if (!adapter.isSolana) {
                    console.log(`Skipping non-Solana wallet: ${wallet.name}`);
                    return;
                }
                const detectedWallet = {
                    wallet,
                    adapter
                };
                this.detectedWalletStandardWallets.set(wallet.name, detectedWallet);
                console.log(`Solana Wallet Standard wallet detected:`, wallet);
                console.log(`  - Features:`, adapter.getFeatures());
            };
            // Get existing wallets
            const existingWallets = get();
            existingWallets.forEach(handleWallet);
            // Listen for new wallets
            const unsubscribe = on('register', (...wallets) => {
                wallets.forEach(handleWallet);
            });
            // Listen for wallet unregistration
            const unsubscribeUnregister = on('unregister', (...wallets) => {
                wallets.forEach(wallet => {
                    this.detectedWalletStandardWallets.delete(wallet.name);
                    console.log(`Wallet Standard wallet unregistered: ${wallet.name}`);
                });
            });
            // Store unsubscribe functions for cleanup if needed
            this.walletStandardUnsubscribe = () => {
                unsubscribe();
                unsubscribeUnregister();
            };
        }
        catch (error) {
            console.log('BridgeParent: Wallet Standard not available or error during detection', error);
        }
    }
    logAllWallets() {
        setTimeout(() => {
            console.log('=== All Detected Wallet Providers (Filtered) ===');
            const totalWallets = this.detectedEip6963Wallets.size +
                this.detectedWalletStandardWallets.size;
            console.log(`Total wallets detected: ${totalWallets}`);
            if (this.detectedEip6963Wallets.size > 0) {
                console.log('\nEIP-6963 Wallets:');
                this.detectedEip6963Wallets.forEach((wallet, uuid) => {
                    console.log(`  - ${wallet.info.name} (UUID: ${uuid}, RDNS: ${wallet.info.rdns})`);
                });
            }
            if (this.detectedWalletStandardWallets.size > 0) {
                console.log('\nSolana Wallet Standard Wallets:');
                this.detectedWalletStandardWallets.forEach((detected, name) => {
                    console.log(`  - ${name} (Version: ${detected.wallet.version}, Chains: ${detected.wallet.chains.join(', ')})`);
                });
            }
            if (totalWallets === 0) {
                console.log('No wallets detected');
            }
            console.log('=====================================');
        }, 1000);
    }
    getDetectedEip6963Wallets() {
        return new Map(this.detectedEip6963Wallets);
    }
    getDetectedWalletStandardWallets() {
        return new Map(this.detectedWalletStandardWallets);
    }
    // Convenience methods for backward compatibility
    getEIP6963Providers() {
        const providers = new Map();
        this.detectedEip6963Wallets.forEach((wallet, uuid) => {
            providers.set(uuid, { info: wallet.info, provider: wallet.provider });
        });
        return providers;
    }
    getWalletStandardWallets() {
        const wallets = new Map();
        this.detectedWalletStandardWallets.forEach((detected, name) => {
            wallets.set(name, detected.wallet);
        });
        return wallets;
    }
    getWalletStandardAdapters() {
        const adapters = new Map();
        this.detectedWalletStandardWallets.forEach((detected, name) => {
            adapters.set(name, detected.adapter);
        });
        return adapters;
    }
    getWalletStandardAdapter(walletName) {
        const detected = this.detectedWalletStandardWallets.get(walletName);
        return detected?.adapter;
    }
    refreshDetection() {
        // Clean up existing wallet standard subscription if exists
        if (this.walletStandardUnsubscribe) {
            this.walletStandardUnsubscribe();
        }
        this.detectedEip6963Wallets.clear();
        this.detectedWalletStandardWallets.clear();
        this.initialize();
    }
}
