// @meshconnect/uwc-bridge-parent/dist/comlink-bridge.js
import * as Comlink from 'comlink';
import { BridgeParent } from './bridge-parent';
class WalletBridge {
    bridgeParent;
    eip6963Wallets = [];
    walletStandardWallets = [];
    constructor() {
        this.bridgeParent = new BridgeParent();
    }
    async initialize() {
        // BridgeParent initializes itself in constructor
        // Just refresh detection to ensure we have latest wallets
        this.bridgeParent.refreshDetection();
        await this.detectWallets();
    }
    async detectWallets() {
        // Get EIP-6963 wallets from BridgeParent
        const eip6963Wallets = this.bridgeParent.getDetectedEip6963Wallets();
        this.eip6963Wallets = Array.from(eip6963Wallets.entries()).map(([uuid, wallet]) => {
            const provider = wallet.provider;
            return {
                uuid,
                name: wallet.info.name,
                info: {
                    uuid,
                    name: wallet.info.name,
                    icon: wallet.info.icon,
                    rdns: wallet.info.rdns
                },
                methods: {
                    request: Comlink.proxy(provider.request.bind(provider)),
                    on: provider.on
                        ? Comlink.proxy(provider.on.bind(provider))
                        : undefined,
                    removeListener: provider.removeListener
                        ? Comlink.proxy(provider.removeListener.bind(provider))
                        : undefined
                }
            };
        });
        // Get Wallet Standard wallets from BridgeParent
        const walletStandardAdapters = this.bridgeParent.getWalletStandardAdapters();
        this.walletStandardWallets = Array.from(walletStandardAdapters.entries())
            .filter(([_, adapter]) => adapter.getSolanaAdapter())
            .map(([name, adapter]) => {
            const solanaAdapter = adapter.getSolanaAdapter();
            return {
                uuid: name.toLowerCase().replace(/\s+/g, '-'),
                name,
                info: {
                    name,
                    chains: ['solana:mainnet'],
                    features: []
                },
                methods: {
                    connect: solanaAdapter?.connect
                        ? Comlink.proxy(solanaAdapter.connect.bind(solanaAdapter))
                        : undefined,
                    signMessage: solanaAdapter?.signMessage
                        ? Comlink.proxy(solanaAdapter.signMessage.bind(solanaAdapter))
                        : undefined,
                    sendTransaction: solanaAdapter?.sendTransaction
                        ? Comlink.proxy(solanaAdapter.sendTransaction.bind(solanaAdapter))
                        : undefined,
                    get name() {
                        return solanaAdapter?.name;
                    },
                    get publicKey() {
                        return solanaAdapter?.publicKey;
                    }
                }
            };
        });
        console.log(`[Comlink Bridge] Found ${this.eip6963Wallets.length} EIP-6963 and ${this.walletStandardWallets.length} Wallet Standard wallets`);
    }
    async getEIP6963Wallets() {
        console.log('[Comlink Bridge] Sending EIP-6963 wallets to child');
        return this.eip6963Wallets;
    }
    async getWalletStandardWallets() {
        console.log('[Comlink Bridge] Sending Wallet Standard wallets to child');
        return this.walletStandardWallets;
    }
    async refreshWallets() {
        console.log('[Comlink Bridge] Refreshing wallets...');
        await this.bridgeParent.refreshDetection();
        await this.detectWallets();
    }
}
/**
 * Simple class to initialize Comlink bridge for an iframe
 */
export class ComlinkBridgeParent {
    bridge = null;
    /**
     * Connect to an iframe and set up the Comlink bridge
     */
    async connectToIframe(iframe) {
        // Wait for iframe to load
        await new Promise(resolve => {
            if (iframe.contentWindow) {
                resolve();
            }
            else {
                iframe.addEventListener('load', () => resolve(), { once: true });
            }
        });
        console.log('[ComlinkBridgeParent] Setting up bridge...');
        // Create and initialize bridge
        this.bridge = new WalletBridge();
        await this.bridge.initialize();
        // Expose bridge to iframe using Comlink
        Comlink.expose(this.bridge, Comlink.windowEndpoint(iframe.contentWindow));
        console.log('[ComlinkBridgeParent] Bridge exposed to iframe');
    }
    /**
     * Refresh wallet detection
     */
    async refreshWallets() {
        if (!this.bridge) {
            throw new Error('Bridge not initialized. Call connectToIframe first.');
        }
        await this.bridge.refreshWallets();
    }
}
