// @meshconnect/uwc-bridge-parent/dist/standard-wallet-adapter.js
/**
 * Adapter for Wallet Standard wallets to provide Solana adapter interface
 */
export class StandardWalletAdapter {
    wallet;
    constructor(wallet) {
        this.wallet = wallet;
    }
    /**
     * Check if this is a Solana wallet
     */
    get isSolana() {
        const chains = this.wallet.chains || [];
        return chains.some(chain => chain.startsWith('solana:'));
    }
    /**
     * Get Solana adapter if available
     */
    getSolanaAdapter() {
        if (!this.isSolana)
            return null;
        // Check for standard Solana features
        const features = this.wallet.features || {};
        // Create adapter-like interface from wallet features
        return {
            name: this.wallet.name,
            publicKey: null,
            connected: false,
            connecting: false,
            readyState: 'Installed',
            // Methods from wallet features
            connect: features['standard:connect'] ?
                features['standard:connect'].connect :
                async () => { console.warn('Connect not supported'); },
            disconnect: features['standard:disconnect'] ?
                features['standard:disconnect'].disconnect :
                async () => { console.warn('Disconnect not supported'); },
            signMessage: features['solana:signMessage'] ?
                features['solana:signMessage'].signMessage :
                undefined,
            signTransaction: features['solana:signTransaction'] ?
                features['solana:signTransaction'].signTransaction :
                undefined,
            signAllTransactions: features['solana:signAllTransactions'] ?
                features['solana:signAllTransactions'].signAllTransactions :
                undefined,
            sendTransaction: features['solana:sendTransaction'] ?
                features['solana:sendTransaction'].sendTransaction :
                undefined
        };
    }
}
