// @meshconnect/uwc-bridge-parent/dist/wallet-bridge-parent.js
import * as Comlink from 'comlink';
import { getWallets } from '@wallet-standard/app';
import { StandardWalletAdapter } from './standard-wallet-adapter';
/**
 * WalletBridgeParent - Manages wallet detection in parent window and exposes to iframe
 */
export class WalletBridgeParent {
    bridge = null;
    /**
     * Initialize bridge connection to iframe
     */
    async initialize(iframe) {
        // Wait for iframe to be ready
        await this.waitForIframe(iframe);
        // Create bridge implementation
        this.bridge = new BridgeImplementation();
        await this.bridge.detectWallets();
        // Expose bridge to iframe via Comlink
        Comlink.expose(this.bridge, Comlink.windowEndpoint(iframe.contentWindow));
        console.log('[WalletBridgeParent] Bridge initialized and exposed to iframe');
    }
    async waitForIframe(iframe) {
        return new Promise(resolve => {
            if (iframe.contentWindow) {
                resolve();
            }
            else {
                iframe.addEventListener('load', () => resolve(), { once: true });
            }
        });
    }
    /**
     * Refresh wallet detection
     */
    async refreshWallets() {
        if (this.bridge) {
            await this.bridge.refreshWallets();
        }
    }
}
/**
 * Bridge implementation that gets exposed via Comlink
 */
class BridgeImplementation {
    eip6963Wallets = [];
    walletStandardWallets = [];
    /**
     * Detect all wallets
     */
    async detectWallets() {
        console.log('[BridgeImplementation] Starting wallet detection...');
        await this.detectEIP6963Wallets();
        await this.detectWalletStandardWallets();
        console.log('[BridgeImplementation] Wallet detection complete:', {
            eip6963: this.eip6963Wallets.length,
            walletStandard: this.walletStandardWallets.length
        });
    }
    /**
     * Detect EIP-6963 wallets and create proxied versions
     */
    async detectEIP6963Wallets() {
        if (typeof window === 'undefined')
            return;
        return new Promise(resolve => {
            const detectedWallets = new Map();
            const handleAnnouncement = (event) => {
                const { info, provider } = event.detail;
                detectedWallets.set(info.uuid, { info, provider });
                console.log(`[BridgeImplementation] EIP-6963 wallet detected: ${info.name}`);
            };
            window.addEventListener('eip6963:announceProvider', handleAnnouncement);
            window.dispatchEvent(new Event('eip6963:requestProvider'));
            // Stop listening after 2 seconds and process wallets
            setTimeout(() => {
                window.removeEventListener('eip6963:announceProvider', handleAnnouncement);
                console.log(`[BridgeImplementation] Processing ${detectedWallets.size} detected EIP-6963 wallets`);
                // Create proxied wallets
                this.eip6963Wallets = [];
                detectedWallets.forEach((wallet, uuid) => {
                    const provider = wallet.provider;
                    // Create a provider object with wrapped methods
                    const proxiedProvider = {
                        // Wrap the request method
                        request: provider.request ?
                            Comlink.proxy(async (args) => provider.request(args)) :
                            undefined,
                        // Wrap event methods
                        on: provider.on ?
                            Comlink.proxy((event, handler) => provider.on(event, handler)) :
                            undefined,
                        removeListener: provider.removeListener ?
                            Comlink.proxy((event, handler) => provider.removeListener(event, handler)) :
                            undefined,
                        // Include properties
                        isMetaMask: provider.isMetaMask,
                        isPhantom: provider.isPhantom,
                        isCoinbaseWallet: provider.isCoinbaseWallet,
                        isTrust: provider.isTrust
                    };
                    this.eip6963Wallets.push({
                        uuid,
                        info: wallet.info,
                        provider: proxiedProvider
                    });
                });
                console.log(`[BridgeImplementation] Processed ${this.eip6963Wallets.length} EIP-6963 wallets`);
                resolve();
            }, 2000);
        });
    }
    /**
     * Detect Wallet Standard wallets and create proxied versions
     */
    async detectWalletStandardWallets() {
        if (typeof window === 'undefined')
            return;
        try {
            const { get } = getWallets();
            const wallets = get();
            this.walletStandardWallets = [];
            for (const wallet of wallets) {
                // Check if it's a Solana wallet
                if (!this.isSolanaWallet(wallet))
                    continue;
                // Create adapter
                const adapter = new StandardWalletAdapter(wallet);
                if (!adapter.isSolana)
                    continue;
                const solanaAdapter = adapter.getSolanaAdapter();
                if (!solanaAdapter)
                    continue;
                // Create Comlink proxies for wallet methods
                const proxiedWallet = {
                    name: wallet.name,
                    icon: wallet.icon,
                    chains: wallet.chains,
                    features: Object.keys(wallet.features)
                };
                // Proxy wallet features if they have methods
                for (const [featureName, feature] of Object.entries(wallet.features)) {
                    if (typeof feature === 'object' && feature !== null) {
                        const proxiedFeature = {};
                        for (const [key, value] of Object.entries(feature)) {
                            if (typeof value === 'function') {
                                proxiedFeature[key] = Comlink.proxy(value.bind(feature));
                            }
                            else {
                                proxiedFeature[key] = value;
                            }
                        }
                        proxiedWallet[featureName] = proxiedFeature;
                    }
                }
                // Create Comlink proxies for adapter methods
                const proxiedAdapter = {
                    name: solanaAdapter.name,
                    publicKey: solanaAdapter.publicKey,
                    connected: solanaAdapter.connected,
                    connecting: solanaAdapter.connecting,
                    readyState: solanaAdapter.readyState
                };
                // Proxy adapter methods
                if (solanaAdapter['connect']) {
                    proxiedAdapter['connect'] = Comlink.proxy(solanaAdapter['connect'].bind(solanaAdapter));
                }
                if (solanaAdapter['disconnect']) {
                    proxiedAdapter['disconnect'] = Comlink.proxy(solanaAdapter['disconnect'].bind(solanaAdapter));
                }
                if (solanaAdapter['signMessage']) {
                    proxiedAdapter['signMessage'] = Comlink.proxy(solanaAdapter['signMessage'].bind(solanaAdapter));
                }
                if (solanaAdapter['signTransaction']) {
                    proxiedAdapter['signTransaction'] = Comlink.proxy(solanaAdapter['signTransaction'].bind(solanaAdapter));
                }
                if (solanaAdapter['signAllTransactions']) {
                    proxiedAdapter['signAllTransactions'] = Comlink.proxy(solanaAdapter['signAllTransactions'].bind(solanaAdapter));
                }
                if (solanaAdapter['sendTransaction']) {
                    proxiedAdapter['sendTransaction'] = Comlink.proxy(solanaAdapter['sendTransaction'].bind(solanaAdapter));
                }
                this.walletStandardWallets.push({
                    name: wallet.name,
                    icon: wallet.icon || '',
                    chains: wallet.chains,
                    features: Object.keys(wallet.features),
                    wallet: proxiedWallet,
                    adapter: proxiedAdapter
                });
                console.log(`[BridgeImplementation] Wallet Standard wallet processed: ${wallet.name}`);
            }
            console.log(`[BridgeImplementation] Processed ${this.walletStandardWallets.length} Wallet Standard wallets`);
        }
        catch (error) {
            console.error('[BridgeImplementation] Error detecting Wallet Standard wallets:', error);
        }
    }
    isSolanaWallet(wallet) {
        const chains = wallet.chains || [];
        return chains.some(chain => chain.startsWith('solana:'));
    }
    /**
     * Get all EIP-6963 wallets with Comlink proxied methods
     */
    async getEIP6963Wallets() {
        console.log(`[BridgeImplementation] Returning ${this.eip6963Wallets.length} EIP-6963 wallets`);
        return this.eip6963Wallets;
    }
    /**
     * Get all Wallet Standard wallets with Comlink proxied methods
     */
    async getWalletStandardWallets() {
        console.log(`[BridgeImplementation] Returning ${this.walletStandardWallets.length} Wallet Standard wallets`);
        return this.walletStandardWallets;
    }
    /**
     * Refresh wallet detection
     */
    async refreshWallets() {
        console.log('[BridgeImplementation] Refreshing wallet detection...');
        await this.detectWallets();
    }
}
