// @meshconnect/uwc-bridge-parent/dist/wallet-bridge.js
import { BridgeParent } from './bridge-parent';
import { PostmateBridge } from './postmate-bridge';
/**
 * Unified class that automatically handles wallet detection and iframe communication
 * Simply initialize with an iframe element and everything is set up automatically
 */
export class WalletBridge {
    bridgeParent;
    postmateBridge;
    iframe = null;
    constructor() {
        // Initialize wallet detection
        this.bridgeParent = new BridgeParent();
        console.log('WalletBridge: Wallet detection initialized');
        // Initialize postmate bridge
        this.postmateBridge = new PostmateBridge(this.bridgeParent);
        console.log('WalletBridge: Bridge initialized');
    }
    /**
     * Initialize the bridge with an iframe element
     * This will automatically set up communication with the child frame
     */
    async connectToIframe(iframe) {
        if (this.iframe) {
            console.warn('WalletBridge: Already connected to an iframe, disconnecting first');
            this.disconnect();
        }
        this.iframe = iframe;
        // Initialize the postmate bridge with the iframe
        await this.postmateBridge.initializeWithIframe(iframe);
        console.log('WalletBridge: Connected to iframe and ready for communication');
    }
    /**
     * Refresh wallet detection
     */
    refreshWallets() {
        this.bridgeParent.refreshDetection();
        console.log('WalletBridge: Wallet detection refreshed');
    }
    /**
     * Get the bridge parent instance (for advanced usage)
     */
    getBridgeParent() {
        return this.bridgeParent;
    }
    /**
     * Get the postmate bridge instance (for advanced usage)
     */
    getPostmateBridge() {
        return this.postmateBridge;
    }
    /**
     * Disconnect from the iframe and clean up
     */
    disconnect() {
        if (this.postmateBridge) {
            this.postmateBridge.destroy();
        }
        this.iframe = null;
        console.log('WalletBridge: Disconnected from iframe');
    }
    /**
     * Destroy the bridge and clean up all resources
     */
    destroy() {
        this.disconnect();
        // Additional cleanup if needed
        console.log('WalletBridge: Destroyed');
    }
}
